File: algorithm/547.number-of-provinces/547.number-of-provinces.java
class Solution {
    // 相似题目: 721账户合并
    public int findCircleNum(int[][] isConnected) {
        return findCircleNumUnionFind(isConnected);
    } 
    private int findCircleNumUnionFind(int[][] isConnected) {
        int n = isConnected.length;
        // build
        int[] parent = new int[n];
        // Map<Integer, Integer> parent = new HashMap<>();
        for(int i = 0; i < n; i++) {
            parent[i] = i;
        }
        // merge
        for(int i = 0; i < n; i++) {
            for(int j = 0; j < n; j++) {
                if(isConnected[i][j] == 1) {
                    merge(parent, i, j);
                }
            }
        }
        // find
        int cnt = 0;
        for(int i = 0; i < n; i++) {
            if(parent[i] == i) {
                cnt++;
            }
        }
        return cnt;
    }
    private void merge(int[] parent, int i, int j) {
        int pi = find(parent, i);
        int pj = find(parent, j);
        parent[pj] = pi;
    }

    private int find(int[] parent, int i) {
        int p = i;
        while(parent[p] != p) {
            p = parent[p];
        }
        return p;
    }
    // Time O(n^2)  Space O(n^2)
    private int findCircleNumDFS(int[][] isConnected) {
        int n = isConnected.length;
        // 标准的无向图 求连通
        boolean[]visited = new boolean[n];
        int cnt = 0;
        // 深度优先遍历
        for(int i = 0; i < n; i++) {
            if(!visited[i]) {
                cnt++;
                dfs(isConnected, i, visited);
            }
        }
        return cnt;
    }

    // def backtrack(路径, 选择列表):
    // graph 选择列表, visited结束条件
    private void dfs(int[][] isConnected, int idx,boolean[]visited) {
        int n = isConnected.length;
        // 退出条件
        if(visited[idx]) {
            return;
        }
        // 操作
        visited[idx] = true;
        // 继续遍历条件
        for(int i = 0; i < n; i++) {
            if(isConnected[idx][i]==1 && !visited[i]) {
                dfs(isConnected, i, visited);
            }
        }
    }
}

File: algorithm/547.number-of-provinces/547.number-of-provinces.py
#
# @lc app=leetcode id=547 lang=python3
#
# [547] Number of Provinces
#

# @lc code=start
# TC union: O(n), isConnected O(1), get O(1)
class UnionFind:
    def __init__(self):
        self.parent = {}
    def union(self, x: int, y: int):
        if not x in self.parent:
            self.parent[x] = x
        if not y in self.parent:
            self.parent[y] = y
        px = self.find(x)
        py = self.find(y)
        if px != py:
            self.parent[px] = py
    def find(self, x: int):
        root = x
        while self.parent[root] != root:
            root = self.parent[root]
        while self.parent[x] != root:
            self.parent[x] = root
            x = self.parent[x]
        return root
        
    def get(self) -> int:
        cnt = set()
        # print(self.parent)
        for x in self.parent:
            cnt.add(self.find(x))
        return len(cnt)

class Solution:
    
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        m, n = len(isConnected), len(isConnected[0])
        self.ans = 0
        visited = [0 for _ in range(0, n)]
        def uinonfind()->int:
            uf = UnionFind()
            for i in range(0, m):
                for j in range(0, n):
                    if(isConnected[i][j] == 1):
                        uf.union(i, j)
            return uf.get()

        # this grid is not a grid of point, they are n cities and grid of link between n cities
        def helper( i: int)->None:
            visited[i] = 1
            for j in range(0, m):
                if(isConnected[i][j] == 1 and visited[j] == 0):
                    helper(j)

        for i in range(0, m):
            if(visited[i] == 0):
                helper(i)
                self.ans += 1
        return self.ans
        # return uinonfind
        
# @lc code=end

